spark/src/main/scala/org/apache/comet/CometSparkSessionExtensions.scala

```diff
@@ -116,143 +113,124 @@ class CometSparkSessionExtensions
-              !CometScanExec.isFileFormatSupported(fileFormat),
-              s"File format $fileFormat is not supported")
-            val info2 = createMessage(
-              !CometScanExec.isSchemaSupported(requiredSchema),
-              s"Schema $requiredSchema is not supported")
-            val info3 = createMessage(
-              !CometScanExec.isSchemaSupported(partitionSchema),
-              s"Partition schema $partitionSchema is not supported")
-            withInfo(scanExec, Seq(info1, info2, info3).flatten.mkString(","))
-            scanExec
         }
       }
     }
+
+    private def transformScan(scanExec: FileSourceScanExec): SparkPlan = {
+
+      def isDynamicPruningFilter(e: Expression): Boolean =
+        e.exists(_.isInstanceOf[PlanExpression[_]])
+
+      if (COMET_DPP_FALLBACK_ENABLED.get() &&
+        scanExec.partitionFilters.exists(isDynamicPruningFilter)) {
+        withInfo(scanExec, "DPP not supported")
+        return scanExec
+      }
+
+      scanExec.relation match {
+        case HadoopFsRelation(_, partitionSchema, _, _, fileFormat, _) =>
+          if (!CometScanExec.isFileFormatSupported(fileFormat)) {
+            withInfo(scanExec, s"fileFormat $fileFormat not supported")
+            return scanExec
+          }
+
+          COMET_NATIVE_SCAN_IMPL.get() match {
+            case SCAN_NATIVE_DATAFUSION =>
+              // TODO we only enable full native scan if COMET_EXEC_ENABLED is enabled
+              // but this is not really what we want .. we currently insert `CometScanExec`
+              // here and then it gets replaced with `CometNativeScanExec` in `CometExecRule`
+              // but that only happens if `COMET_EXEC_ENABLED` is enabled
+              if (!COMET_EXEC_ENABLED.get()) {
+                withInfo(
+                  scanExec,
+                  s"Native scan not enabled when ${COMET_EXEC_ENABLED.key} is not enabled")
+                return scanExec
+              }
+              if (!CometNativeScanExec.isSchemaSupported(scanExec.requiredSchema)) {
+                withInfo(scanExec, s"requiredSchema ${scanExec.requiredSchema} not supported")
+                return scanExec
+              }
+              if (!CometNativeScanExec.isSchemaSupported(partitionSchema)) {
+                withInfo(scanExec, s"partitionSchema $partitionSchema not supported")
+                return scanExec
+              }
+              CometScanExec(scanExec, session)
+
+            case SCAN_NATIVE_COMET | SCAN_NATIVE_ICEBERG_COMPAT =>
+              if (!CometScanExec.isSchemaSupported(scanExec.requiredSchema)) {
+                withInfo(scanExec, s"requiredSchema ${scanExec.requiredSchema} not supported")
+                return scanExec
+              }
+              if (!CometScanExec.isSchemaSupported(partitionSchema)) {
+                withInfo(scanExec, s"partitionSchema $partitionSchema not supported")
+                return scanExec
+              }
```

Comment:
We used to include multiple reasons previously, but this change only return one of them.